nodestream/pipeline/transformers/transformer.py

```diff
@@ -54,83 +57,65 @@ def __init__(
-                else:
-                    remaining_pending_tasks.append(task)
-
-            pending_tasks[:] = remaining_pending_tasks
-            if tasks_drained:
-                self.logger.debug(
-                    "Drained %s completed tasks, %s pending tasks remain",
-                    tasks_drained,
-                    len(pending_tasks),
-                )
-
-        async def drain_completed_tasks_async():
-            for result in drain_completed_tasks():
-                yield result
-            await self.yield_processor()
-
-        async for record in record_stream:
-            if record is Flush:
-                # Flush the pending tasks, then yield the flush.
-                # In order to fully respect the Flush,
-                # we need to wait for all pending tasks to complete.
-                # If we're still waiting we can yield the processor.
-                while pending_tasks:
-                    async for result in drain_completed_tasks_async():
-                        yield result
-                yield record
+    async def drain_completed_tasks(self):
+        tasks_drained = 0
+        remaining_pending_tasks = []
+        for task in self.pending_tasks:
+            if task.done():
+                yield task.result()
+                tasks_drained += 1
             else:
-                # Submit the work to the thread pool (only if we have capacity)
-                # If we don't have capacity, yield completed tasks.
-                # Once there is nothing remaining to yield from these tasks, yield the processor until capacity is available.
-                submitted = False
-                while not submitted:
-                    if len(pending_tasks) < self.maximum_pending_tasks:
-                        task = self.thread_pool.submit(self.do_work_on_record, record)
-                        pending_tasks.append(task)
-                        submitted = True
-                    async for result in drain_completed_tasks_async():
-                        yield result
-
-        # After we've finished enqueuing records, we need to drain all tasks.
-        # Items yielded from this loop are the final results of the step.
-        self.logger.debug("Finished enqueuing records, draining pending tasks")
-        while pending_tasks:
-            async for result in drain_completed_tasks_async():
-                yield result
+                remaining_pending_tasks.append(task)
+
+        await self.yield_processor()
+
+        self.pending_tasks[:] = remaining_pending_tasks
+        if tasks_drained:
+            self.logger.debug(
+                "Drained %s completed tasks, %s pending tasks remain",
+                tasks_drained,
+                len(self.pending_tasks),
+            )
+
+    async def process_record(self, record: Any, _) -> Any:
+        if record is Flush:
+            async for outstanding in self.drain_completed_tasks():
+                yield outstanding
```

Comment:
Outstanding case for a Flush being passed as a record. In the unit test this could just be added as one of the elements in the input generator.